external-crates/move/move-compiler/src/hlir/translate.rs

```diff
@@ -583,782 +597,1088 @@ fn type_(context: &Context, sp!(loc, ty_): N::Type) -> H::Type {
-    use T::LValue_ as L;
-    match bind_ {
-        L::Ignore => (),
-        L::Var { var: v, ty, .. } => {
-            let st = single_type(context, *ty.clone());
-            context.bind_local(*v, st)
+fn tail_block(
+    context: &mut Context,
+    block: &mut Block,
+    expected_type: Option<&H::Type>,
+    mut seq: T::Sequence,
+) -> Option<H::Exp> {
+    use T::SequenceItem_ as S;
+    let last_exp = seq.pop_back();
+    if statement_block(context, block, seq) {
+        if DEBUG_PRINT {
+            println!("divergent statements in tail block");
         }
-        L::Unpack(_, _, _, fields) | L::BorrowUnpack(_, _, _, _, fields) => fields
-            .iter()
-            .for_each(|(_, _, (_, (_, b)))| declare_bind(context, b)),
+        return None;
+    }
+    match last_exp {
+        None => None,
+        Some(sp!(_, S::Seq(last))) => tail(context, block, expected_type, *last),
+        Some(_) => panic!("ICE last sequence item should be an exp"),
     }
 }
 
-fn assign_command(
-    context: &mut Context,
-    result: &mut Block,
-    loc: Loc,
-    sp!(_, assigns): T::LValueList,
-    rvalue: H::Exp,
-) {
-    use H::{Command_ as C, Statement_ as S};
-    let mut lvalues = vec![];
-    let mut after = Block::new();
-    for (idx, a) in assigns.into_iter().enumerate() {
-        let a_ty = rvalue.ty.value.type_at_index(idx);
-        let (ls, mut af) = assign(context, a, a_ty);
+// -------------------------------------------------------------------------------------------------
+// Value Position
+// -------------------------------------------------------------------------------------------------
 
-        lvalues.push(ls);
-        after.append(&mut af);
+fn value(
+    context: &mut Context,
+    block: &mut Block,
+    expected_type: Option<&H::Type>,
+    e: T::Exp,
+) -> Option<H::Exp> {
+    // we pull outthese cases because it's easier to process them without destructuring `e` first.
+    if is_statement(&e) {
+        let result = if is_unit_statement(&e) {
+            Some(unit_exp(e.exp.loc))
+        } else {
+            None
+        };
+        statement(context, block, e);
+        return result;
+    } else if is_binop(&e) {
+        let out_type = type_(context, e.ty.clone());
+        return process_binops(context, block, out_type, e);
     }
-    result.push_back(sp(loc, S::Command(sp(loc, C::Assign(lvalues, rvalue)))));
-    result.append(&mut after);
-}
 
-fn assign(
-    context: &mut Context,
-    sp!(loc, ta_): T::LValue,
-    rvalue_ty: &H::SingleType,
-) -> (H::LValue, Block) {
-    use H::{LValue_ as L, UnannotatedExp_ as E};
-    use T::LValue_ as A;
-    let mut after = Block::new();
-    let l_ = match ta_ {
-        A::Ignore => L::Ignore,
-        A::Var { var: v, ty: st, .. } => {
-            L::Var(translate_var(v), Box::new(single_type(context, *st)))
+    use H::{Command_ as C, Statement_ as S, UnannotatedExp_ as HE};
+    use T::UnannotatedExp_ as E;
```

Comment:
Super minor nit: but stylistically I feel as though having all of the `use` statements at the top of the function is nice if possible (so lifting these and the `use`s in the match branches to the top of this function).